src/ooni/http_invalid_request_line_impl.hpp

```diff
@@ -5,75 +5,107 @@
 
 
   public:
-    HTTPInvalidRequestLineImpl(Settings options_) : HTTPTestImpl(options_) {
+    HTTPInvalidRequestLineImpl(Settings options_) : TCPTestImpl(options_) {
         test_name = "http_invalid_request_line";
         test_version = "0.0.1";
     };
 
+    void send_receive_invalid_request_line(http::Url backend_url,
+            std::string request_line,
+            std::function<void()> &&cb) {
+        connect({{"host", backend_url.address}, {"port", std::to_string(backend_url.port)}},
+                [this, cb, request_line](Var<net::Transport> txp) {
+
+            Var<std::string> received_data(new std::string);
+            txp->on_data([this, received_data](net::Buffer data) {
+                logger.debug("http_invalid_request_line: on_data");
+                *received_data += data.read();
+                logger.debug("%s", received_data->c_str());
+            });
+            txp->write(request_line);
+
+            // We assume to have received all the data after a timeout of 5
+            // seconds.
+            poller->call_later(timeout,
+                    [this, cb, received_data, request_line, txp]() {
+                if (*received_data != request_line) {
+                    logger.info("Tampering detected!");
+                    logger.info("%s != %s", received_data->c_str(), request_line.c_str());
+                    entry["tampering"] = true;
+                } else if (entry["tampering"] == nullptr){
+                    logger.info("Tampering not detected.");
+                    entry["tampering"] = false;
+                }
+                entry["sent"].push_back(request_line);
+                entry["received"].push_back(*received_data);
+                txp->close([this, cb](){
+                    cb();
+                });
+            });
+        });
+    };
+
     void main(Settings options, std::function<void(json)> &&cb) {
+        entry["tampering"] = nullptr;
+
+        http::Url backend_url = mk::http::parse_url(options["backend"]);
 
-        auto handle_response = [this, cb](Error, http::Response &&) {
+        auto handle_response = [this, cb]() {
             tests_run += 1;
-            if (tests_run == 3) {
+            if (tests_run == 4) {
                 cb(entry);
             }
-            // XXX we currently don't set the tampering key, because this test
-            // speaks to a TCP Echo helper, hence the response will not be valid
-            // HTTP.
         };
 
-        http::Headers headers;
         // test_random_invalid_method
         // randomSTR(4) + " / HTTP/1.1\n\r"
-        request(
-            {
-                {"url", options["backend"]},
-                {"method", mk::random_str_uppercase(4)},
-                {"http_version", "HTTP/1.1"},
-            },
-            headers, "", handle_response);
+        std::string test_random_invalid_method(mk::random_str_uppercase(4));
+        test_random_invalid_method += " / HTTP/1.1\n\r";
```

Comment:
Actually this is a bug in ooniprobe and the relative specification for the test.

For consistencies sake I have implemented this in the same way of how it's specified in ooni-probe (https://github.com/TheTorProject/ooni-probe/blob/master/ooni/nettests/manipulation/http_invalid_request_line.py#L75) and in ooni-spec (https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-007-http-invalid-request-line.md), but I think that eventually they should all be updated.

When they are updated they should probably also include a blank line at the end of them.

I will file a ticket about this, since otherwise I will forget.